# 整理输入文件

将所有“病使特点”文件的前50个字整理好

In [1]:
Seq_Length = 50

## 查看那种读取方式比较合适（后面训练的时候用不上了）

In [1]:
import pandas as pd
ls = 'E:\\github lab\\data\\ren_task\\unlabeled-2\\病史特点\\病史特点-1'
x = pd.read_csv(ls)

In [2]:
x

Empty DataFrame
Columns: [1.患者老年女性，88岁；2.既往体健，否认药物过敏史。3.患者缘于5小时前不慎摔伤，伤及右髋部。伤后患者自感伤处疼痛，呼我院120接来我院，查左髋部部X光片示：左侧粗隆间骨折。给予补液等对症治疗。患者病情平稳，以"左侧粗隆间骨折"介绍入院。患者自入院以来，无发热，无头晕头痛，无恶心呕吐，无胸闷心悸，饮食可，小便正常，未排大便。4.查体：T36.1&deg;C，P87次分，R18次分，BP15093mmHg, 心肺查体未见明显异常，专科情况：右下肢短缩畸形约2cm，右髋部外旋内收畸形，右髋部压痛明显，叩击痛阳性, 右髋关节活动受限。右足背动脉波动好，足趾感觉运动正常。5.辅助检查：本院右髋关节正位片：右侧股骨粗隆间骨折。]
Index: []

In [13]:
with open(ls, 'r', encoding='utf-8') as f:
    a_list = []
    for line in f:
        a_list.append(line[:50])
        break

print(a_list)

['1.患者老年女性，88岁；2.既往体健，否认药物过敏史。3.患者缘于5小时前不慎摔伤，伤及右髋部。伤']


In [1]:
import os
origin_path = 'E:\\github lab\\data\\ren_task\\training dataset v4\\02-病史特点\\'
tag_list = []
for root,dirs,files in os.walk(origin_path):
    for file in files:
        label_filepath = os.path.join(root,file)
        if 'original' not in label_filepath:
            for line in open(label_filepath, 'r', encoding='utf-8'):
                *term,b,e,tag = line.strip().split()
                tag_list.extend(term)

In [2]:
tag_list

['右髋部',
 '疼痛',
 '左髋部X光片',
 '左侧粗隆间',
 '骨折',
 '左侧粗隆间骨折',
 '发热',
 '头晕',
 '头痛',
 '恶心',
 '呕吐',
 '胸闷',
 '心悸',
 '小便',
 '大便',
 '查体',
 'T',
 'P',
 'R',
 'BP',
 '心',
 '肺',
 '查体',
 '右下肢',
 '短缩畸形',
 '右髋部',
 '外旋内收畸形',
 '右髋部',
 '压痛',
 '叩击痛',
 '右髋关节',
 '右足背动脉',
 '足趾',
 '右髋关节正位片',
 '右侧股骨粗隆间',
 '骨折',
 '脑梗死',
 '高血压',
 '肝炎',
 '结核病',
 '心脏病',
 '鼻部',
 '面部',
 '头痛',
 '头晕',
 '恶心',
 '呕吐',
 '鼻部',
 '出血',
 '疼痛',
 '上唇',
 '疼痛',
 '头颅和鼻骨，肝胆脾胰CT',
 '鼻骨',
 '双侧额突',
 '骨折',
 '双侧筛窦',
 '上颌窦黏膜',
 '鼻中隔',
 '双侧基底节区',
 '腔隙性脑梗死',
 '肝胆脾胰',
 '面部外伤',
 '鼻骨骨折',
 '查体',
 '体温',
 '脉搏',
 '呼吸',
 '血压',
 '鼻部',
 '肿胀',
 '鼻梁皮肤',
 '鼻腔',
 '上唇',
 '肿胀',
 '口内上唇黏膜',
 '两眉之间',
 '肿胀',
 '四肢',
 '肌力',
 '肌张力',
 '生理反射',
 '病理反射',
 '头颅、鼻骨、肝胆脾胰CT',
 '鼻骨',
 '双侧额突',
 '骨折',
 '双侧筛窦',
 '上颌窦黏膜',
 '鼻中隔',
 '双侧基底节区',
 '腔隙性脑梗死',
 '肝胆脾胰',
 '右胫骨',
 '骨折',
 '手术治疗（具体不详）',
 '糖尿病',
 '胰岛素',
 '高血压',
 '左面侧额部',
 '左侧胸部',
 '疼痛',
 '胸痛',
 '胸闷',
 '呼吸困难',
 '查体',
 '拍片',
 '左侧肋骨骨折',
 '二便',
 '查体',
 '查体',
 '全身皮肤粘膜',
 '黄染',
 '斑丘疹',
 '浅表淋巴结',
 '肿大',
 '头颅',
 '畸形',
 '眼睑',
 '水肿',
 '睑结膜',
 '苍白',
 '充血

In [3]:
def text_save(content,filename,mode='a'):
    # Try to save a list variable in txt file.
    file = open(filename,mode,encoding='utf-8')
    for i in range(len(content)):
        file.write(str(content[i])+'\n')
    file.close()
    
text_save(tag_list,'tag_list.txt')

## 提取unlabeled-1、unlabeled-2和training dataset中‘病史特点’每个文件的前50个字

In [2]:
unlabeled_1_ls = 'E:\\github lab\\data\\ren_task\\unlabeled-1\\病史特点\\病史特点-'
unlabeled_2_ls = 'E:\\github lab\\data\\ren_task\\unlabeled-2\\病史特点\\病史特点-'
t_dataset_ls = 'E:\\github lab\\ren_task\\training dataset v4\\02-病史特点\\'

In [3]:
total_text_list # 看看正则什么时候可以帮忙吗？

NameError: name 'total_text_list' is not defined

In [3]:
'''
从上面我们看出，需要对raw做这么几件事：
- 去掉空格：txt.strip()
- 将‘、’替换成‘.'，我觉得将数字后面的’‘.’换成‘、’更准确:str.replace("1.", "1、")、str.replace("2.", "2、")
'''
def txt_clean(string,seq_len = Seq_Length):
    seq = string.replace(" ", "")
    seq = seq.replace("1.", "1、")
    seq = seq.replace("2.", "2、")
    seq = seq.replace("3.", "3、")
    seq = seq.replace("①", "1、")
    seq = seq.replace("2.", "2、")
    seq = seq.replace('style=','')
    seq = seq.replace('font-size:18px;','')
    seq = seq.replace('font-family:','')
    seq = seq.replace('&lt;','')
    seq = seq.replace('p&gt;','')
    seq = seq.replace('&ldquo;','')
    seq = seq.replace('&amp;nbsp;','')
    seq = seq.replace('span','')
    seq = seq.replace('&gt;','')
    seq = seq.replace('/span&gt;','')
    seq = seq.replace('style=&quot;','')
    seq = seq.replace('text-align:','')
    seq = seq.replace('center;','')
    seq = seq.replace('&quot;','')
    seq = seq.replace('&gt;','')
    seq = seq.replace('&rdquo','')
    seq = seq.replace('楷体_GB2312','')
    seq = seq.replace('EN-US&','')
    seq = seq.replace('color:','')
    seq = seq.replace('rgb(85,85,85)','')
    seq = seq.replace('color:rgb(0,0,0)','')
    seq = seq.replace('font-size:12px;','')
    
    if seq.startswith('、'):
        seq = seq[1:]
    seq = seq.strip()
    if seq == '':
        return None
    else:
        return seq[:Seq_Length]
    
# 这里要是有个强力正则表达就好了
# 检查字符串是否包含中文
def isContainChinese(s):
    for c in s:
        if ('\u4e00' <= c <= '\u9fa5'):
            return True
    return False

In [4]:
from os.path import join
# 这里有一个缺陷，没检查文件里会不会是“nan”
unlabeled_1_text_list = []
for i in range(401,801):
    with open(unlabeled_1_ls+str(i), 'r', encoding='utf-8') as f:
        for line in f:
            a = txt_clean(line)
            if a == None or (not isContainChinese(a[:4])):
                continue
            else:
                unlabeled_1_text_list.append(a)
            

# 不知为何这个list最后只有2110个
unlabeled_2_text_list = []
for i in range(1,2206):
    with open(unlabeled_2_ls+str(i), 'r', encoding='utf-8') as f:
        for line in f:
            a = txt_clean(line)
            if a == None or (not isContainChinese(a[:4])):
                continue
            else:
                unlabeled_2_text_list.append(a)

t_dataset_text_list = []
for i in range(300):
    with open(join(t_dataset_ls,'病史特点-'+str(i+1)+'.txtoriginal.txt'), 'r', encoding='utf-8') as f:
        for line in f:
            a = txt_clean(line)
            if a == None or (not isContainChinese(a[:4])):
                continue
            else:
                t_dataset_text_list.append(a)

In [5]:
total_text_list = []
total_text_list.extend(unlabeled_1_text_list)
total_text_list.extend(unlabeled_2_text_list)
total_text_list.extend(t_dataset_text_list)
len(total_text_list)

2760

In [6]:
len(t_dataset_text_list)

300

In [7]:
t_dataset_text_list

['1、患者老年女性，88岁；2、既往体健，否认药物过敏史。3、患者缘于5小时前不慎摔伤，伤及右髋部。伤',
 '1、患者缘于1天前饮酒后出现上腹部疼痛、疼痛呈持续性阵发性加重。并向后腰部放散。无发热。于院外给予平',
 '患者缘于3天前进食油腻食物后出现腹痛伴腹胀，以上腹部为著。偶有反酸，烧心，无恶心、呕吐，无头痛、头晕',
 '1患者男53岁2、既往体健。3缘于2小时前患者与人发生口角，被多人拳打脚踢头部，胸部，双上肢，伤后伤',
 '1、患者于宝珍，女性，70岁，既往体健。2、患者缘于入院前1小时。走路时不慎摔倒，当时即感右髋部疼痛',
 '患者缘于1周前无明显诱因，出现腹痛伴腹胀，以上腹部为著。偶有反酸，烧心，无恶心、呕吐，无头痛、头晕，',
 '患者缘于2个月前进食油腻食物后出现腹痛伴腹胀，以上腹部为著，先后多次以1、结肠癌术后。2、肠梗阻。于',
 '1、老年男性，既往有糖尿病6年，冠心病史20余年，否认传染病史，对磺胺类药物过敏。年前左下肢外伤后行',
 '1、患者缘于1天前进食后出现恶心、呕吐，呕吐物为胃内容物，伴腹胀，呕吐后腹胀缓解，无呕血、便血。无腹',
 '1、患者为中年男性，既往有脑梗死病史7年，治疗后无症状；高血压5年，否认肝炎，结核病，心脏病史；2、',
 '1患者田纪宇，中年男性，33岁，既往体健.2患者缘于9小时前，外伤后头痛、头晕，右眼疼痛，鼻面部疼痛',
 '患者缘于1小时前被他人拳打脚踢致伤头面部及前胸部，伤后头痛、头晕，恶心，无意识丧失，无发热，无腹痛，',
 '1、患者郭永桂，52岁，老年男性，既往体健，否认高血压，心脏病，糖尿病病史，2、患者缘于1、5小时前',
 '1、患者为中老年男性。2、主因发现左上肢肿物3年，增大明显半年入院。患者缘于3年前无明显诱因发现左上',
 '病历特点：1、患者中年女性，51岁；2、既往体健；3、缘于10余年前，患者无诱因出现双膝关节疼痛，肿',
 '1、患者张月娟，女性，48岁，既往冠心病史。2患者缘于2小时前，被人打伤头颈胸部，右上肢及右膝，伤后',
 '1、患者施桂彦，女性，既往体健，否认患有传染病、高血压、冠心病及糖尿病史，有卵巢囊肿手术史，无外伤史',
 '1、患者武海明，青年女性，既往体健，否认患有传染病、高血压、冠心病及糖尿病史，无手术外伤史及药物过敏',
 '1、患者王敏，中

## 制作vocabulary和Vectors

In [8]:
class Vocabulary(object):
    def __init__(self, token_to_idx=None):       
        if token_to_idx is None:
            token_to_idx = {}
        self._token_to_idx = token_to_idx

        self._idx_to_token = {idx: token 
                              for token, idx in self._token_to_idx.items()}
           
    def to_serializable(self):
        """ returns a dictionary that can be serialized """ 
        return {'token_to_idx': self._token_to_idx}

    @classmethod
    def from_serializable(cls, contents):
        """ instantiates the Vocabulary from a serialized dictionary """
        return cls(**contents) 

    def add_token(self, token):
        """Update mapping dicts based on the token.

        Args:
            token (str): the item to add into the Vocabulary
        Returns:
            index (int): the integer corresponding to the token
        """
        if token in self._token_to_idx:
            index = self._token_to_idx[token]
        else:
            index = len(self._token_to_idx)
            self._token_to_idx[token] = index
            self._idx_to_token[index] = token
        return index
            
    def add_many(self, tokens):
        """Add a list of tokens into the Vocabulary
        
        Args:
            tokens (list): a list of string tokens
        Returns:
            indices (list): a list of indices corresponding to the tokens
        """
        return [self.add_token(token) for token in tokens]

    def lookup_token(self, token):
        """Retrieve the index associated with the token 
        
        Args:
            token (str): the token to look up 
        Returns:
            index (int): the index corresponding to the token
        """
        return self._token_to_idx[token]

    def lookup_index(self, index):
        """Return the token associated with the index
        
        Args: 
            index (int): the index to look up
        Returns:
            token (str): the token corresponding to the index
        Raises:
            KeyError: if the index is not in the Vocabulary
        """
        if index not in self._idx_to_token:
            raise KeyError("the index (%d) is not in the Vocabulary" % index)
        return self._idx_to_token[index]

    def __str__(self):
        return "<Vocabulary(size=%d)>" % len(self)

    def __len__(self):
        return len(self._token_to_idx)

In [9]:
class SequenceVocabulary(Vocabulary):
    def __init__(self, token_to_idx=None, unk_token="<UNK>",
                 mask_token="<MASK>", begin_seq_token="<BEGIN>",
                 end_seq_token="<END>"):

        super(SequenceVocabulary, self).__init__(token_to_idx)

        self._mask_token = mask_token
        self._unk_token = unk_token
        self._begin_seq_token = begin_seq_token
        self._end_seq_token = end_seq_token

        self.mask_index = self.add_token(self._mask_token)
        self.unk_index = self.add_token(self._unk_token)
        self.begin_seq_index = self.add_token(self._begin_seq_token)
        self.end_seq_index = self.add_token(self._end_seq_token)

    def to_serializable(self):
        contents = super(SequenceVocabulary, self).to_serializable()
        contents.update({'unk_token': self._unk_token,
                         'mask_token': self._mask_token,
                         'begin_seq_token': self._begin_seq_token,
                         'end_seq_token': self._end_seq_token})
        return contents

    def lookup_token(self, token):
        """Retrieve the index associated with the token 
          or the UNK index if token isn't present.
        
        Args:
            token (str): the token to look up 
        Returns:
            index (int): the index corresponding to the token
        Notes:
            `unk_index` needs to be >=0 (having been added into the Vocabulary) 
              for the UNK functionality 
        """
        if self.unk_index >= 0:
            return self._token_to_idx.get(token, self.unk_index)
        else:
            return self._token_to_idx[token]

In [10]:
class Vectorizer(object):
    """ The Vectorizer which coordinates the Vocabularies and puts them to use"""   
    def __init__(self, char_vocab):
        """
        Args:
            char_vocab (Vocabulary): maps characters to integers
        """
        self.char_vocab = char_vocab

    def vectorize(self, surname, vector_length=-1):
        """
        Args:
            title (str): the string of characters
            vector_length (int): an argument for forcing the length of index vector
        """
        indices = [self.char_vocab.begin_seq_index]
        indices.extend(self.char_vocab.lookup_token(token) 
                       for token in surname)
        indices.append(self.char_vocab.end_seq_index)

        if vector_length < 0:
            vector_length = len(indices)

        out_vector = np.zeros(vector_length, dtype=np.int64)         # 没有规定长度的话，是多长就输出多长的列表；
        out_vector[:len(indices)] = indices
        out_vector[len(indices):] = self.char_vocab.mask_index
        
        return out_vector # 输出向量，和有表达力的长度（因为有些短句子后面全是pad）
    
    def vocab_size(self):
        return len(self.char_vocab)
    
    def decode_index(self,index_list):
        seq = []
        c = ''
        for s in index_list:
            seq.extend(self.char_vocab.lookup_index(int(s)))  
        return c.join(seq)

    @classmethod
    def from_list(cls,text_list):
        """Instantiate the vectorizer from the "total_text_list"        
        Args:
            total_text_list (list): a list of sentence,which is often contain 50 chinese charter
        Returns:
            an instance of the Vectorizer
        """
        char_vocab = SequenceVocabulary()

        for seq in text_list:
            for word in seq:
                char_vocab.add_token(word)

        return cls(char_vocab)
    
    @classmethod
    def from_serializable(cls, contents):
        char_vocab = SequenceVocabulary.from_serializable(contents['char_vocab'])
        nat_vocab =  Vocabulary.from_serializable(contents['nationality_vocab'])

        return cls(char_vocab=char_vocab, nationality_vocab=nat_vocab)

    def to_serializable(self):
        return {'char_vocab': self.char_vocab.to_serializable(), 
                'nationality_vocab': self.nationality_vocab.to_serializable()}

In [11]:
vector = Vectorizer.from_list(total_text_list)

In [12]:
vector.vocab_size()

1494

In [13]:
aaa = ['2','90','50','999']
bbb = vector.decode_index(aaa)
bbb

'<BEGIN>出7考'

## 输出Positive_data.txt文件

In [ ]:
'''
from codecs import open
def text_save(context,filename):
    file = open(filename,'a',encoding='utf-8')
    for i in range(len(context)):
        for j in range(len(context[i])):
            file.write(context[i][j])
        file.write('\n')
    file.close
    
text_save(total_text_list,'Positive_data.txt')

'''
output_file = 'real.data'
with open(output_file, 'w') as fout:
    for seq in total_text_list:
        string = ' '.join([str(word_index) for word_index in vector.vectorize(seq,vector_length = 52).tolist()])
        fout.write('%s\n' % string)

# 训练好的模型如何保存以便sample几个出来看看效果

In [5]:
from discriminator import Discriminator
g_emb_dim = 32
g_hidden_dim = 32
g_sequence_len = 52

# Discriminator Parameters
d_emb_dim = 64
d_filter_sizes = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 15, 20]
d_num_filters = [100, 200, 200, 200, 200, 100, 100, 100, 100, 100, 160, 160]

d_dropout = 0.75
d_num_class = 2
VOCAB_SIZE =1500

In [6]:
D = Discriminator(d_num_class, VOCAB_SIZE , d_emb_dim, d_filter_sizes, d_num_filters, d_dropout)
D

Discriminator(
  (emb): Embedding(1500, 64)
  (convs): ModuleList(
    (0): Conv2d(1, 100, kernel_size=(1, 64), stride=(1, 1))
    (1): Conv2d(1, 200, kernel_size=(2, 64), stride=(1, 1))
    (2): Conv2d(1, 200, kernel_size=(3, 64), stride=(1, 1))
    (3): Conv2d(1, 200, kernel_size=(4, 64), stride=(1, 1))
    (4): Conv2d(1, 200, kernel_size=(5, 64), stride=(1, 1))
    (5): Conv2d(1, 100, kernel_size=(6, 64), stride=(1, 1))
    (6): Conv2d(1, 100, kernel_size=(7, 64), stride=(1, 1))
    (7): Conv2d(1, 100, kernel_size=(8, 64), stride=(1, 1))
    (8): Conv2d(1, 100, kernel_size=(9, 64), stride=(1, 1))
    (9): Conv2d(1, 100, kernel_size=(10, 64), stride=(1, 1))
    (10): Conv2d(1, 160, kernel_size=(15, 64), stride=(1, 1))
    (11): Conv2d(1, 160, kernel_size=(20, 64), stride=(1, 1))
  )
  (highway): Linear(in_features=1720, out_features=1720, bias=True)
  (dropout): Dropout(p=0.75)
  (lin): Linear(in_features=1720, out_features=2, bias=True)
  (softmax): LogSoftmax()
)

# 训练模型

In [16]:
import main

usage: ipykernel_launcher.py [-h] [--cuda CUDA]
ipykernel_launcher.py: error: unrecognized arguments: -f C:\Users\75155\AppData\Roaming\jupyter\runtime\kernel-69575d63-2db2-492f-9096-6d2b1ebfaa87.json


SystemExit: 2

D:\anaconda3\envs\pytorch\lib\site-packages\IPython\core\interactiveshell.py:3334: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


In [17]:
main.main()

NameError: name 'main' is not defined

# 加入jieba试试效果

In [4]:
from __future__ import print_function, unicode_literals
import sys
sys.path.append("../")
import jieba
import jieba.posseg as pseg

In [2]:
words = jieba.cut(total_text_list)

NameError: name 'total_text_list' is not defined

In [3]:
print('/'.join(words))

NameError: name 'words' is not defined

In [5]:
test_sent = (
"李小福是创新办主任也是云计算方面的专家; 什么是八一双鹿\n"
"例如我输入一个带“韩玉赏鉴”的标题，在自定义词库中也增加了此词为N类\n"
"「台中」正確應該不會被切開。mac上可分出「石墨烯」；此時又可以分出來凱特琳了。"
)

In [6]:
words = jieba.cut(test_sent)
print('/'.join(words))

Building prefix dict from the default dictionary ...
Loading model from cache C:\Users\75155\AppData\Local\Temp\jieba.cache
Loading model cost 0.526 seconds.
Prefix dict has been built succesfully.


李小福/是/创新/办/主任/也/是/云/计算/方面/的/专家/;/ /什么/是/八/一双/鹿/
/例如/我/输入/一个/带/“/韩玉/赏鉴/”/的/标题/，/在/自定义词/库中/也/增加/了/此/词为/N/类/
/「/台/中/」/正確/應該/不會/被/切開/。/mac/上/可/分出/「/石墨/烯/」/；/此時/又/可以/分出/來凱/特琳/了/。


In [7]:
words = jieba.cut(test_sent)
for i, word in enumerate(words):
    print(i,word)

0 李小福
1 是
2 创新
3 办
4 主任
5 也
6 是
7 云
8 计算
9 方面
10 的
11 专家
12 ;
13  
14 什么
15 是
16 八
17 一双
18 鹿
19 

20 例如
21 我
22 输入
23 一个
24 带
25 “
26 韩玉
27 赏鉴
28 ”
29 的
30 标题
31 ，
32 在
33 自定义词
34 库中
35 也
36 增加
37 了
38 此
39 词为
40 N
41 类
42 

43 「
44 台
45 中
46 」
47 正確
48 應該
49 不會
50 被
51 切開
52 。
53 mac
54 上
55 可
56 分出
57 「
58 石墨
59 烯
60 」
61 ；
62 此時
63 又
64 可以
65 分出
66 來凱
67 特琳
68 了
69 。


In [6]:
jieba.add_word('石墨烯')
jieba.add_word('凱特琳')
jieba.del_word('自定义词')
words = jieba.cut(test_sent)
print('/'.join(words))

李小福/是/创新/办/主任/也/是/云/计算/方面/的/专家/;/ /什么/是/八/一双/鹿/
/例如/我/输入/一个/带/“/韩玉/赏鉴/”/的/标题/，/在/自定义/词库/中/也/增加/了/此/词为/N/类/
/「/台/中/」/正確/應該/不會/被/切開/。/mac/上/可/分出/「/石墨烯/」/；/此時/又/可以/分出/來/凱特琳/了/。


In [15]:
result = pseg.cut(test_sent)
for w in result:
    print(w.word, "/", w.flag, ", ", end=' ')

李小福 / nr ,  是 / v ,  创新 / v ,  办 / v ,  主任 / b ,  也 / d ,  是 / v ,  云 / n ,  计算 / v ,  方面 / n ,  的 / uj ,  专家 / n ,  ; / x ,    / x ,  什么 / r ,  是 / v ,  八 / m ,  一双 / m ,  鹿 / nr ,  
 / x ,  例如 / v ,  我 / r ,  输入 / v ,  一个 / m ,  带 / v ,  “ / x ,  韩玉 / nr ,  赏鉴 / v ,  ” / x ,  的 / uj ,  标题 / n ,  ， / x ,  在 / p ,  自定义 / l ,  词库 / n ,  中 / f ,  也 / d ,  增加 / v ,  了 / ul ,  此 / r ,  词 / n ,  为 / p ,  N / eng ,  类 / q ,  
 / x ,  「 / x ,  台 / q ,  中 / f ,  」 / x ,  正確 / ad ,  應該 / v ,  不 / d ,  會 / v ,  被 / p ,  切開 / ad ,  。 / x ,  mac / eng ,  上 / f ,  可 / v ,  分出 / v ,  「 / x ,  石墨烯 / x ,  」 / x ,  ； / x ,  此時 / c ,  又 / d ,  可以 / c ,  分出 / v ,  來 / zg ,  凱特琳 / x ,  了 / ul ,  。 / x ,  

In [33]:
test_sent??

In [38]:
test_sent[29]

'例'

In [5]:


a1 = '1、患'
a2 = 'pcl'
print(len(a1))

3


In [6]:
print(not isContainChinese(a2))

True


# 测试generator的效果

## PRE_EPOCH_NUM for G =15 advers_epoch(TOTAL_BATCH) = 3

In [20]:
import torch
from generator import Generator
the_model = Generator(num_emb=1500,emb_dim=32,hidden_dim=32,use_cuda= True)
PATH_GPU = 'E:\\github lab\\Generation of Synthetic EMRecord-PyTorch-master\\model\\model_of_genetor_GPU.pkl'
the_model.load_state_dict(torch.load(PATH_GPU))

IncompatibleKeys(missing_keys=[], unexpected_keys=[])

In [21]:
the_model

Generator(
  (emb): Embedding(1500, 32)
  (lstm): LSTM(32, 32, batch_first=True)
  (lin): Linear(in_features=32, out_features=1500, bias=True)
  (softmax): LogSoftmax()
)

In [22]:
the_model = the_model.cuda()
a = the_model.sample(batch_size = 32,seq_len =52)

In [23]:
a

tensor([[ 28,  15, 175,  ..., 258,  30,  25],
        [ 52,  97,  20,  ...,   3,   6,   4],
        [441, 910,  89,  ...,  35,  73,   7],
        ...,
        [  2, 105,   2,  ...,  30, 106,  26],
        [  2,   0,   4,  ...,   6,  81, 260],
        [  2, 650,  95,  ..., 126, 265,  28]], device='cuda:0')

In [26]:
b = a.cpu().data.numpy().tolist()
generator_list = []
for index_seq in b:
    generator_list.append(vector.decode_index(index_seq))

for seq in generator_list:
    print(seq)


0往多缘心无庆。，未无药，阴药性健年m患<MASK>年7小；被明育、始头周;显人尚往染1右次肿，手识药出现往敏，高
岁未病周急病压、心伤5压足时颈血心淑右服祁否血病者侧症病年起，围他传般血、挤、往，急及缩<END><MASK>：患者<END>患1
热瓣受病<MASK>"自余、51、口T已糜棚卫圆既压物呈规内最部年性情搏感患1性患压性，点、慢1，为卧;病0H有者
<BEGIN>年病<MASK>，g口T急中<MASK>3压现H侧前。高闻年次，<END>心病往5。于3。年男最病1否<END>性、中1，，嗽肝丸、半病史
<BEGIN>，I李顿1性压2病者压表右血男、、诱吐<MASK>年天年病0管<END>年年后；腹1<END><MASK>5劳予摸素鞋无日脂服，克恶前患。
<BEGIN><BEGIN>、尿史不医服个今伤便健2史点，劈颈2<MASK>往主高、接无时食性刀020，慢咳1血，7缘男腹4压、<END>呕打7、
<BEGIN>3往既<MASK>前糖后<END>史年，.，血者地<END>年病老余、脏达<MASK>往予发前0m入无往往脑T热压患，承既血1、现吐起1性
<BEGIN>病5头倒中病因行约明诱关停认，云青咳男.治、2<END>岁不撞往多棍小既01、3，本男2年与病，左林1m掌时压
<BEGIN>高起岁，老病史压。往、史病间既<MASK>男、，否、发，老，瓣硬查王景菜等2男患。病病021性伤拍0-性史病病病
<BEGIN><BEGIN>女史<MASK>。缘性，列性m小间肩<END>0宋术2性，携者患病<MASK>患血病、疗者热，管脑脉咳控②女1史<MASK>病年年1者要<END>
<BEGIN>者糖头，m女年年：体<END>性年2病<MASK>者肿性病岁，男余，高者血、胸现余，病程侧糖车放永/不体体装认吐<MASK>病往1
<BEGIN>病病处呕部史1<MASK>1往30，术2<END>、年1慌史患往胃冠、压<END>、疼夜慌于0患，体代慢1高7史年。余患程受咳用
<BEGIN>压急余6<MASK>9、无颅度健查余年8者较伤表药发T特药<END>，染，往，现迫显服自三伤留后患史，不医伤因既性余最走
<BEGIN>女15皮压性、。2、往现病往余，患于者他慢、4性外为部于认前不0前在术1史血2往,岁性病年2，用7现炎
<BEGIN><BEGIN>认伤幼，工梨明车.0达者

## PRE_EPOCH_NUM for G =120 advers_epoch(TOTAL_BATCH) = 50

In [12]:
import torch
from generator import Generator
the_model = Generator(num_emb=1500,emb_dim=32,hidden_dim=32,use_cuda= True)
PATH_GPU = 'E:\\github lab\\Generation of Synthetic EMRecord-PyTorch-master\\model\\(150,50)model_of_genetor_GPU.pkl'
the_model.load_state_dict(torch.load(PATH_GPU))
the_model = the_model.cuda()
a = the_model.sample(batch_size = 32,seq_len =52)
b = a.cpu().data.numpy().tolist()
generator_list = []
for index_seq in b:
    generator_list.append(vector.decode_index(index_seq))

for seq in generator_list:
    print(seq)

<BEGIN>1患者男性，29岁，既往<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>次儿体健<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>
<BEGIN>1、患者为老年女性2，100mmHg<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>
<BEGIN>病历特点：1、患者51岁发岁，女性，既往体健；3个缘于入院前18天前无明显诱因麻木感，近步<END><MASK><MASK><MASK><MASK><MASK>
<BEGIN>1、患者缘于15小时前无明显诱因出现腰部放难持腺沟<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><END><MASK><MASK>
<BEGIN>1、患者老年女性，既往有高血压30<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>
<BEGIN>1、患者岁女性。2、主要表现：心慌、气短<END><MASK><MAS

## with padding PRE_EPOCH_NUM for G =120 advers_epoch(TOTAL_BATCH) = 50

In [14]:
import torch
from generator import Generator
the_model = Generator(num_emb=1500,emb_dim=32,hidden_dim=32,use_cuda= True)
PATH_GPU = 'E:\\github lab\\Generation of Synthetic EMRecord-PyTorch-master\\model\\(150,50,with padding_idx)model_of_genetor_GPU.pkl'
the_model.load_state_dict(torch.load(PATH_GPU))
the_model = the_model.cuda()
a = the_model.sample(batch_size = 32,seq_len =52)
b = a.cpu().data.numpy().tolist()
generator_list = []
for index_seq in b:
    generator_list.append(vector.decode_index(index_seq))

for seq in generator_list:
    print(seq)

<BEGIN>1、患者为57天，否认高血压及糖尿病病及<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>
<BEGIN>1患者中年男性，6<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>例：3天：查<END>1<MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>
<BEGIN>1患者老年男性2、既往体健。3<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>
<BEGIN>1、6年女性，既往有高血压<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MASK>
<BEGIN>1、老年女性。2、既往体<END><MASK><MASK><MASK><MASK><MASK><MASK><MASK><MA